[1 - 3단계 방탈출 사용자 예약] 투다(박찬양) 미션 제출합니다.

praisebak opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
안녕하세요 아서! 레벨1때 뵙고 또 뵙게되어서 너무 반갑습니다.
이전에 비해 스스로 많이 성장했다고 느끼는데 아서가 체감하실 수 있도록 이번에도 열심히 해보겠습니다!
👏 

수정된 파일들

https://github.com/woowacourse/spring-roomescape-member/pull/219/files/3c408be42ddf315761eb933c31941ebd0152011b..d5fee702722d778088ac5a5a39b30b4931c89657
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4280
## 얼마나 객체지향적으로 구현했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [ ] 4
- [x] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
완성을한 코드를 살펴보았을때
문제점을 찾아내지 못해서 한번 과감하게 5번 체크해보았습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작
제 코드에서 시작하였지만, service 의존 방식에서 페어의 repository 페어 방식으로 수정하고 시작하였습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

## 예약가능한 시간 가져오기 기능 - 컨트롤러 api 명명 고민
  - reservations/available-times vs times/available-times
  - reservations/available-times
    - 예약을 하기 위해 가능한 시간들을 가져오는것이라고 판단하여 전자를 선택하였습니다.
- 그런데 추후 살펴보니 자원을 기반으로 url을 설정해야한다는 점에서, 후자가 맞다고 생각이 듭니다.
시간 관계상 이번에는 수정하지못하였으나 추후에 restful한 명명으로 수정해보겠습니다.

## e2e 테스트 작성후 느낀점
- 테스트에서 중요한 부분은 비즈니스로직이라고 생각합니다.
- 그러면서도 서비스 로직과 컨트롤러,레포지토리가 가지고있는 의존성들과도 잘 동작하는지가 검증되지 않는게 꺼름칙하여 새 e2e 테스트를 작성해보았습니다.
  - 작성해보니 검증이 잘되어서 안정적이라는 생각보다 번거롭다는 생각이 더 많이 들었습니다.
    - 컨트롤러는 한번 만들어지면 수정될 일이 잘 없고, 수정을하여도 해당 컨트롤러만 수정한뒤 - 실제로 돌려보는것으로 충분하는 생각이 들었습니다.
  - 따라서 다음에는 서비스 db를 모킹한 단위 테스트와, 서비스에 실제 db를 사용하는 통합 테스트로 작성해볼 생각입니다.
    - 컨트롤러까지 검증하기 위해 테스트 작성 시간, 매번 테스트가 돌아갈때의 시간 소모가 되는것은 별로라고 생각하였습니다 
위 생각이 너무 안일한것일지 아서의 의견도 궁금합니다!

## Service간 의존하지 않는 이유
- 이번에는 각 Service가 여러 Repository만을 가지는 구조를 채택하였습니다.
  - Service 방식이 재활용측면에서 뛰어나다고 생각하지만, 이번 미션에서는 순환 참조 문제가 발생하였습니다.
    - 이를 위해 새로운 서비스를 만드는 방법도 있지만, 마땅한 추상화가 떠오르지않았고 중간에 새 계층이 불필요하다 생각하였습니다.
  - 현재 미션의 작은 규모에 대해서는 Repository 참조 방식도 봐줄만하다고 생각하여 사용하였습니다.

## 도메인이 비즈니스로직을 최대한 가지게해보자
(entity를 단순히 영속성을 기대하는 도메인 객체라고 생각합니다.) 

- 서비스 계층이 나오면서 많은 비즈니스로직을 가지게 되었습니다.
  - 기존에 객체지향에서배웠던 도메인이 하는일은 단순한 데이터를 가지기만하는 것이 불과했습니다.
- 이 문제를 해결하기위해서 어색하지 않은 선에서 도메인이 비즈니스로직을 최대한 많이 가지게하는 것을 목표로 하였습니다
- 이를 위해 ThemeRanking,ReservationSlotTimes라는 id를 가지지않는 도메인 객체를 구현하였습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

praisebak committed:
chore: 초기 코드 추가

praisebak committed:
chore: 초기 코드 추가

praisebak committed:
chore: 기존 코드 추가

praisebak committed:
docs: README 추가

praisebak committed:
fix: 시간 추가시 null이면 500 -> 400 예외 발생하도록 수정

praisebak committed:
feat: 예약이 존재하는 시간 삭제 시도 시, 예외 발생 구현

- 서비스 의존 방식에서 레포지토리 방식으로 변경
순환참조 문제 발생으로 인해 ApplicationService을 고려하였으나 새로운 계층 추가로 인한 복잡성이 더 커진다 생각하여 Repository 의존 방식으로 구현하였습니다.

praisebak committed:
feat: 중복 시간 설정시 예외 발생 구현

praisebak committed:
feat: 중복된 예약 시도 시, 예외 발생

praisebak committed:
refactor: 검증 순서 수정

praisebak committed:
feat: 예외 핸들러 추가

praisebak committed:
feat: 테마 생성 및 조회기능 추가

praisebak committed:
merge

praisebak committed:
feat: 테마 삭제 기능 추가

praisebak committed:
feat: 테마 CRUD api 구현

praisebak committed:
feat: 예약 테마 기능 구현

praisebak committed:
feat: tmp

praisebak committed:
chore: 페어 백업

praisebak committed:
feat: 유저 예약 기능 구현

praisebak committed:
feat: 랭킹 기능 구현

praisebak committed:
test: 서비스 테스트 커버리지 향상

praisebak committed:
refactor: 서비스 의존 -> repository 의존 방식으로 바뀌며 불필요해진 트랜잭션 제거

praisebak committed:
refactor: 랭킹 기능 서비스 메서드 -> 객체로 분리

praisebak committed:
refactor: service validate -> 일부분 reservation 객체에 좀 더 책임을 주도록 수정

praisebak committed:
refactor: 가독성 향상

praisebak committed:
docs: 리드미 추가

praisebak committed:
test: 통합테스트 추가

praisebak committed:
fix: 통합테스트 동작하도록 수정

praisebak committed:
chore: 불필요한 파일 삭제

praisebak commented:
현재 e2e 테스트 부분에서 데이터베이스 관련해서 초기화 과정에서 실패하는 문제가 있는것같습니다
(1회에는 성공, 2회에는 실패,그 다음 실행에는 성공)

이부분은 다음 리뷰에 수정하여 반영해보도록하겠습니다!

praisebak commented:
@coderabbitai 도움을 주십시오.

Hyunta commented on src/main/java/roomescape/controller/MainPageController.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping("/")
+public class MainPageController {
+
+    @GetMapping("reservation")
```
RequestMapping 으로 `/` 를 관리해주시는 이유가 있나요?
`/reservation` 으로 써보면 어떤가요?

praisebak replied:
그렇네요 MainPage라고해서 꼭 RequstMapping이 "/"일 필요는없을것같습니다 다음과 같이 수정해보았습니다

praisebak replied:
[refactor: requestMapping 명시 수정](https://github.com/woowacourse/spring-roomescape-member/pull/219/commits/de452c62d809c8252c7bff4f90fa9f7193fd735f)

Hyunta replied:
제가 말씀드렸던 것 `@RequestMapping` 자체가 필요 없겠다는 의미이긴 했습니다.
어차피 메서드가 하나뿐이라서 어떻게 구현해도 큰 차이는 없을 것 같습니다.

Hyunta commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -32,20 +39,49 @@ public ResponseEntity<List<ReservationResponseDto>> reservations() {
 
     @PostMapping
-    public ResponseEntity<Void> addReservations(@RequestBody @Valid AddReservationDto newReservationDto) {
+    public ResponseEntity<ReservationResponseDto> addReservations(
+            @RequestBody @Valid AddReservationDto newReservationDto) {
```
`@Valid` 가 미션 진행하는데 꼭 필요했나요?

praisebak replied:
validation 기능을 사용하면 직접 로직을 짜지 않고도 유효성 검증을 할 수 있다는 점이 매력적이었습니다.
저는 Valid를 컨트롤러에서 받는 dto에 대해서만 걸어주었는데요,
위와 같이 구현하면 Junit만으로는 테스트하지 못하는 문제가 발생합니다.

그러나 현재 컨트롤러에 대해서 Junit만을 이용한 단위 테스트는 필요성을 느끼지 못하였습니다.
저라면 컨트롤러는 스프링 컨텍스트가 동반된 혹은 MockMvc 등을 도입하여 테스트를 작성할것이기 때문에 순수자바로 유지할 필요가 없다고 생각하여 해당 기능을 사용하였습니다.

Hyunta commented on src/main/java/roomescape/dto/AddReservationDto.java:
```diff
@@ -2,16 +2,16 @@
 
-public record AddReservationDto(@NotNull @NotBlank(message = "이름이 비어있을 수 없습니다.") String name,
-                                @FutureOrPresent(message = "날짜는 현재보다 미래여야합니다.") LocalDate date, Long timeId) {
+public record AddReservationDto(@NotBlank(message = "이름이 비어있을 수 없습니다.") String name,
+                                @FutureOrPresent(message = "날짜는 현재보다 미래여야합니다.") LocalDate date, Long timeId,
```
`Validation` 의존성이 없었더라면 어떻게 관리하셨을 것인가요?
지금은 스프링을 학습하는 단계이니 해당 의존성을 제거하고 진행해보시면 어떤가요?

praisebak replied:
`record` 대신에 `class`로 바꾸어서 생성자에 검증 로직을 넣을것같습니다.
메시지를 추가하기 위해 예외 클래스에 메시지를 주입하는 식으로 구현할것같습니다.
 
혹시 학습용도라면 서비스나 도메인단에서는 `Validation`을 사용하지않고 있어서 
경험은 해본것같은데
이부분은 생략을 해도 괜찮을까요..?

Hyunta replied:
네, 이 부분은 그냥 넘어가시죠.

Hyunta commented on src/main/java/roomescape/domain/ReservationSlot.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.LocalTime;
+import java.util.Objects;
+
+public class ReservationSlot {
+    private final Long id;
+    private final LocalTime time;
+    private final boolean reserved;
+
+    public ReservationSlot(Long id, LocalTime time, boolean reserved) {
+        this.id = id;
+        this.time = time;
+        this.reserved = reserved;
+    }
```
`ReservationSlot` 어떤 역할을 하는 도메인인가요?

praisebak replied:
예약이 이미 되었는지를 관리하는 도메인입니다.
지금은 하는일이 많지 않지만, 향후에 `예약여부에 따른 ` 기능이 추가된다면
해당 도메인을 통해 구현하라는 의도로 작성하였습니다.

Hyunta commented on src/main/java/roomescape/domain/ThemeRanking.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.Map;
+
+public class ThemeRanking {
```
랭킹을 왜 별도의 도메인으로 만들어주셨을까요?

praisebak replied:
서비스의 비즈니스로직 수행 부담을 어느정도 도메인에게 덜어주려고 하였습니다.
추후에 랭킹 기능이 다른 서비스에서 사용된다고하면 서비스를 의존하기보다 해당 도메인을 의존하면 되는 장점도 있구요.
(서비스를 의존하게되는것 보다는 작은 단위로 의존하는것이어서 이점이 있다 생각합니다)

그 결과 `랭킹`과 연관된 메시지는 해당 객체에서 처리하기를 기대하고 해당 객체를 구현하였습니다.

## 서비스에서 다 하면 안될까요?
스스로 위 질문을 던져보았습니다.

Service계층의 역할은 모델,DB 의존성을 통해 데이터를 가지고와서 비즈니스로직을 처리하는 것이라고 생각하였습니다.
그 중 최대한 비즈니스로직을 처리하는 부분은 최대한 위와 같이 도메인을 통해 해소하고자 하였습니다.

Hyunta replied:
좋은 시도라고 생각됩니다.
이번에는 서비스 로직들을 별도의 도메인으로 분리해서 가져가는 시도를 해보시고,
구현해보시면서 느껴지는 장단점을 정리해보시면 좋은 경험이 될 것 같습니다.

Hyunta commented on src/main/java/roomescape/repository/JdbcThemeRepository.java:
```diff
@@ -0,0 +1,68 @@
+    public long add(Theme theme) {
+        String sql = "insert into theme (name,description,thumbnail) values(?,?,?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        int update = jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
+            ps.setString(1, theme.getName());
+            ps.setString(2, theme.getDescription());
+            ps.setString(3, theme.getThumbnail());
+            return ps;
+        }, keyHolder);
```
변수로 관리할 필요가 없지 않나요?

praisebak replied:
그렇군요 사용하지 않는 변수는 삭제하겠습니다!

Hyunta commented on README.md:
```diff
@@ -0,0 +1,32 @@
+# README
+
```
> 작성해보니 검증이 잘되어서 안정적이라는 생각보다 번거롭다는 생각이 더 많이 들었습니다.
컨트롤러는 한번 만들어지면 수정될 일이 잘 없고, 수정을하여도 해당 컨트롤러만 수정한뒤 - 실제로 돌려보는것으로 충분하는 생각이 들었습니다.
따라서 다음에는 서비스 db를 모킹한 단위 테스트와, 서비스에 실제 db를 사용하는 통합 테스트로 작성해볼 생각입니다.
컨트롤러까지 검증하기 위해 테스트 작성 시간, 매번 테스트가 돌아갈때의 시간 소모가 되는것은 별로라고 생각하였습니다
위 생각이 너무 안일한것일지 아서의 의견도 궁금합니다!

충분히 그렇게 생각해볼 수 있다고 생각합니다.
그런데 만약 동료가 테스트를 안해본 상태로 PR을 올렸다면 어땠을까요?
직접 테스트했는지 물어보고 확인하는 것보다는 작성되어 있는 테스트를 자동으로 돌려보고 테스트하는 것이 안정적이지 않을까요?

praisebak replied:
맞습니다 테스트가 많으면 많을수록 안정적이라고 생각합니다.
그래서 테스트를 안하는것은 안정성이 필요하지않다고 말하는것과 같습니다.

아서가 말씀해주신것과 같이 동료가 테스트를 잘했는지를 증명하는것 또한 테스트코드기도 하구요.
그렇지만 모든 api에 대해서 1대1로 작성하는것은 조금 꺼려지는것같습니다.
단순히 컨트롤러가 서비스에서 호출하는것을 리턴만한다면 이부분은 생략가능하다고 볼것같습니다.
 
결국 작성은 해야겠지만, 최대한 효율적으로 작성하는 방법을 찾아봐야할것같습니다!

Hyunta commented on src/main/java/roomescape/domain/ThemeRanking.java:
```diff
@@ -0,0 +1,30 @@
+        this.themeRankCount = calculateThemeRankCount;
+    }
+
+    public List<Theme> getAscendingRanking() {
+        return themeRankCount.entrySet().stream()
+                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
+                .map(Map.Entry::getKey)
+                .limit(RANKING_SIZE)
+                .toList();
+    }
```
이런 기능에 대한 테스트는 없어도 괜찮나요?

praisebak replied:
있어야하는데 작성하지않았습니다!
수정하도록하겠습니다!

Hyunta reviewed (changes requested):
안녕하세요 투다,
두번째 미션으로 만나게 된 아서입니다.
몇가지 커멘트 남겼는데 확인 후 재요청 부탁드립니다.
궁금하거나 모호한 부분은 DM으로 알려주세요.

praisebak committed:
fix: 통합 테스트 실패하던 문제 해결

DirtiesContext는 스프링이 캐시중인 컨텍스트를 지운다   -> delete from 으로 테이블의 데이터를 삭제한다 -> id는 여전히 증가되어있다
애초에 db에서 생성하는 id에 의존하는 테스트로 작성하지 않는것이 좋았다.

기존 코드를 유지하기위해 DirtiesContext를 사용한다면 AfterEach로 매번 테이블을 날려야한다

praisebak committed:
refactor: 만약 인기 테마에 대한 새로운 날짜 요구사항이 생기는 상황을 대비하여 좀 더 유동적인 코드로 수정

praisebak committed:
refactor: requestMapping 명시 수정

praisebak commented:
통합테스트 실패 문제를 해결하였습니다~

근본적인 원인은 id에 의존하는 테스트 작성이었습니다.
이번에는 테스트 코드를 수정하지 않고 수정할 수 있는 방법을 채택해보았습니다.

[fix: 통합 테스트 실패하던 문제 해결](https://github.com/woowacourse/spring-roomescape-member/pull/219/commits/3128de01ff0fc39650ff49f0640899cbd0d035b6)

praisebak committed:
refactor: 사용하지 않는 변수 삭제

praisebak committed:
fix: 잘못된 문법 수정

praisebak committed:
test: theme ranking test 추가

praisebak committed:
test: reservationSlot 테스트 추가

praisebak committed:
test: reservationSlot 테스트 추가

praisebak committed:
Merge branch 'step1' of https://github.com/praisebak/spring-roomescape-member into step1

Hyunta reviewed (approved):
안녕하세요 투다,
피드백 반영 잘 해주셨네요.
몇가지 추가 커멘트 남겼는데 다음 미션 하시면서 확인해주세요.
모호하거나, 추가로 궁금한 부분이 있으면 DM 주세요.
이번 단계는 여기서 머지하도록 하겠습니다.
